### PR TITLE
[docs] Add int4wo + 2:4 sparsity example to TorchAO README

### PR DESCRIPTION
**Pull request by jcaip**
# What does this PR do?

This PR updates the documentation for TorchAO to include int4wo + 2:4 sparsity

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by jcaip** [comment]
cc @sayakpaul @SunMarc @MekkCyber

Just some updates for torchao int4wo + 2:4 sparsity docs, please feel free to reassign if you aren't the right person.

File: docs/source/en/quantization/torchao.md [hunk 1]
```diff
@@ -38,6 +38,7 @@ torchao supports the [quantization techniques](https://github.com/pytorch/ao/blo
 - A8W8 Int8 Dynamic Quantization
 - A16W8 Int8 Weight Only Quantization
 - A16W4 Int4 Weight Only Quantization
+- A16W4 Int4 Weight Only Quantization + 2:4 Sparsity
```
Comment (sayakpaul) [inline docs/source/en/quantization/torchao.md:41]: Do we have to let the users know what TorchAO version they should be using for this?

**Review by sayakpaul — APPROVED** [review]

File: docs/source/en/quantization/torchao.md [hunk 2]
```diff
@@ -147,6 +148,37 @@ print(tokenizer.decode(output[0], skip_special_tokens=True))
...
 </hfoptions>
 
+</hfoption>
+<hfoption id="int4-weight-only-24sparse">
+
+```py
+import torch
+from transformers import TorchAoConfig, AutoModelForCausalLM, AutoTokenizer
+from torchao.quantization import Int4WeightOnlyConfig
+from torchao.dtypes import MarlinSparseLayout
```
Comment (MekkCyber) [inline docs/source/en/quantization/torchao.md:158]: If time permits, adding a test for this would be very helpful

File: docs/source/en/quantization/torchao.md [hunk 3]
```diff
@@ -147,6 +148,37 @@ print(tokenizer.decode(output[0], skip_special_tokens=True))
...
+quant_config = Int4WeightOnlyConfig(layout=MarlinSparseLayout())
+quantization_config = TorchAoConfig(quant_type=quant_config)
+
+# Load and quantize the model with sparsity. A sparse checkpoint is needed to accelerate without accuraccy loss
+quantized_model = AutoModelForCausalLM.from_pretrained(
+    "RedHatAI/Sparse-Llama-3.1-8B-2of4",
+    torch_dtype=torch.float16,
+    device_map="cuda",
+    quantization_config=quantization_config
+)
```
Comment (MekkCyber) [inline docs/source/en/quantization/torchao.md:169]: I think it would be great to explain how do we know if the model is a MarlinSparseLayout or an other layout

**Review by MekkCyber — APPROVED** [review]
Thanks a lot for adding this 🤗

File: docs/source/en/quantization/torchao.md [hunk 4]
```diff
@@ -147,6 +148,37 @@ print(tokenizer.decode(output[0], skip_special_tokens=True))
...
+from transformers import TorchAoConfig, AutoModelForCausalLM, AutoTokenizer
+from torchao.quantization import Int4WeightOnlyConfig
+from torchao.dtypes import MarlinSparseLayout
+
+quant_config = Int4WeightOnlyConfig(layout=MarlinSparseLayout())
+quantization_config = TorchAoConfig(quant_type=quant_config)
+
+# Load and quantize the model with sparsity. A sparse checkpoint is needed to accelerate without accuraccy loss
+quantized_model = AutoModelForCausalLM.from_pretrained(
+    "RedHatAI/Sparse-Llama-3.1-8B-2of4",
```
Comment (SunMarc) [inline docs/source/en/quantization/torchao.md:165]: how can we create a sparse checkpoint ?

**Review by SunMarc — COMMENTED** [review]
Thanks ! Left a comment

File: docs/source/en/quantization/torchao.md (hunk 4 above)
Comment (jcaip) [reply docs/source/en/quantization/torchao.md:165]: This is a open area of research, it's not something currently supported by torchao. This checkpoint was made by a combination of https://arxiv.org/abs/2310.06927 and https://arxiv.org/abs/2301.00774

File: docs/source/en/quantization/torchao.md (hunk 1 above)
Comment (jcaip) [reply docs/source/en/quantization/torchao.md:41]: It's been present in torchao for a while, just getting around to updating the docs now.

File: docs/source/en/quantization/torchao.md (hunk 3 above)
Comment (jcaip) [reply docs/source/en/quantization/torchao.md:169]: You can print the model to see the Layout used

```
LlamaForCausalLM(
  (model): LlamaModel(
    (embed_tokens): Embedding(128256, 4096)
    (layers): ModuleList(
      (0-31): 32 x LlamaDecoderLayer(
        (self_attn): LlamaAttention(
          (q_proj): Linear(in_features=4096, out_features=4096, weight=AffineQuantizedTensor(shape=torch.Size([4096, 4096]), block_size=(1, 128), device=cuda:0, _layout=MarlinSparseLayout(), tensor_impl_dtype=torch.int32, quant_min=0, quant_max=15))
          (k_proj): Linear(in_features=4096, out_features=1024, weight=AffineQuantizedTensor(shape=torch.Size([1024, 4096]), block_size=(1, 128), device=cuda:0, _layout=MarlinSparseLayout(), tensor_impl_dtype=torch.int32, quant_min=0, quant_max=15))
          (v_proj): Linear(in_features=4096, out_features=1024, weight=AffineQuantizedTensor(shape=torch.Size([1024, 4096]), block_size=(1, 128), device=cuda:0, _layout=MarlinSparseLayout(), tensor_impl_dtype=torch.int32, quant_min=0, quant_max=15))
          (o_proj): Linear(in_features=4096, out_features=4096, weight=AffineQuantizedTensor(shape=torch.Size([4096, 4096]), block_size=(1, 128), device=cuda:0, _layout=MarlinSparseLayout(), tensor_impl_dtype=torch.int32, quant_min=0, quant_max=15))
        )
        (mlp): LlamaMLP(
          (gate_proj): Linear(in_features=4096, out_features=14336, weight=AffineQuantizedTensor(shape=torch.Size([14336, 4096]), block_size=(1, 128), device=cuda:0, _layout=MarlinSparseLayout(), tensor_impl_dtype=torch.int32, quant_min=0, quant_max=15))
          (up_proj): Linear(in_features=4096, out_features=14336, weight=AffineQuantizedTensor(shape=torch.Size([14336, 4096]), block_size=(1, 128), device=cuda:0, _layout=MarlinSparseLayout(), tensor_impl_dtype=torch.int32, quant_min=0, quant_max=15))
          (down_proj): Linear(in_features=14336, out_features=4096, weight=AffineQuantizedTensor(shape=torch.Size([4096, 14336]), block_size=(1, 128), device=cuda:0, _layout=MarlinSparseLayout(), tensor_impl_dtype=torch.int32, quant_min=0, quant_max=15))
          (act_fn): SiLU()
        )
        (input_layernorm): LlamaRMSNorm((4096,), eps=1e-05)
        (post_attention_layernorm): LlamaRMSNorm((4096,), eps=1e-05)
      )
    )
    (norm): LlamaRMSNorm((4096,), eps=1e-05)
    (rotary_emb): LlamaRotaryEmbedding()
  )
  (lm_head): Linear(in_features=4096, out_features=128256, bias=False)
)
```

**Review by stevhliu — APPROVED** [review]
LGTM, thanks!
